components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java

```diff
@@ -41,32 +42,40 @@ public PulsarProducer(PulsarEndpoint pulsarEndpoint) {
     @Override
     public void process(final Exchange exchange) throws Exception {
         final Message message = exchange.getIn();
-        final String topic = pulsarEndpoint.getTopic();
-        final String producerName = topic + "-" + Thread.currentThread().getId();
-
-        final Map properties = message.getHeader(PulsarMessageHeaders.PROPERTIES, Map.class);
-
-        final ProducerBuilder<byte[]> producerBuilder = pulsarEndpoint
-            .getPulsarClient()
-            .newProducer()
-            .producerName(producerName)
-            .topic(topic);
-
-        if (properties != null && !properties.isEmpty()) {
-            producerBuilder.properties(properties);
+        if(producer == null) {
+            createProducer();
```

Comment:
Why not create the producer in the doStart method?

Reply:
yes, probably it's better.

Reply:
I was following the pattern that is used by the camel-JMS component to lazily create the producer.

Reply:
I'll move it to doStart now

Reply:
Ah okay yeah so camel-jms is a bit more dynamic as its producers can be affinited to a request/reply pattern with temporary queues and all sorts of complexities

Reply:
If you want to offer that you can connect to pulsar on demand, then i suggest to add a new option on the endpoint which the end user can then turn on. And maybe have the option on the component level too, so you can set it generally

Reply:
As some ppl want to fail fast if connection issues, and others want it on-demand. And btw for pulsar does it have any kind of connection pooling concept with automatic re-connection or some sorts of that?

Reply:
When a pulsar client connects to pulsar there is built in support for multiple pulsar brokers and if one goes down automatically connecting to another. It does this using a zookeeper cluster that is used to keep track of all the meta data around pulsar. I understand that it gracefully handles connection issues.

Reply:
Have moved the creation of the Pulsar producer into the doStart method.